File: 77.组合.swift
/*
 * @lc app=leetcode.cn id=77 lang=swift
 *
 * [77] 组合
 */

// @lc code=start
//1. 弄清楚 N 和 N-1 之间的关系，这里要求的组合是不重复的。返回条件是k==1的时候
//2. 注意小标， startIndex从1开始，而不是从0， 判断条件是tartIndex <= n - count + 1， 用特殊值试一下。例如4 2
class Solution {
    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        var result = [[Int]]()
        
        func combine(prefix: [Int], start: Int, count: Int){
            guard count > 1 else {
                for i in start...n {
                    var value = prefix
                    value.append(i)
                    result.append(value)
                }
                return
            }
            
            var startIndex = start
            while startIndex <= n - count + 1{
                var nextPrefix = prefix
                nextPrefix.append(startIndex)
                combine(prefix: nextPrefix, start: startIndex + 1, count: count - 1)
                startIndex += 1
            }
            
        }

        combine(prefix: [], start: 1, count: k)
        return result
    }
}
// @lc code=end

File: 77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start

from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if k > n:
            return []
        
        result = []
        self.helper([], n, k, result)
        return result
        
    def helper(self, prefix:List[int], n: int, k: int, result: List[List[int]]):
        if k == 0:
            result.append(prefix)
            return
        if n == 0:
            return 
        
        copy2 = prefix.copy()
        copy2.append(n)
        self.helper(copy2, n - 1, k-1, result)
        
        copy1 = prefix.copy()
        self.helper(copy1, n - 1, k, result)
        
result = Solution().combine(4, 2)        
print(result)
# @lc code=end

